src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs

```diff
@@ -632,6 +632,30 @@ private static string GenerateClassName(TypeDeclarationSyntax typeDeclaration)
 
                 bool onMostDerivedType = true;
 
+#if ROSLYN4_8_OR_GREATER
+                IEnumerable<IMethodSymbol> primaryConstructors = classType.InstanceConstructors
+                    .Where(ic => ic.DeclaringSyntaxReferences
+                        .Any(ds => ds.GetSyntax() is ClassDeclarationSyntax));
+
+                foreach (IMethodSymbol primaryConstructor in primaryConstructors)
```

Comment:
You should also test scenarios like this:

```cs
class Test(ILogger logger)
{
    private readonly ILogger _logger = logger;
}
```

and ensure that you're using the right logger for that scenario.

Reply:
@333fred is it possible to lookup the fields first and then if didn't find any try to look up the primary constructors? 

Also, for detecting the primary constructor, should check `constructor.Body == null && constructor.ExpressionBody == null`?

Reply:
> private readonly ILogger _logger = logger;

good test as we don't support private fields.

Reply:
> @333fred is it possible to lookup the fields first and then if didn't find any try to look up the primary constructors?

Sure, that would be possible. Talking with Aleksey, that's likely the approach you want to take in general. However, I'm going to strongly recommend that you explicitly consider these cases and lay out, in comments and documentation, the exact rules you follow. IE, what happens when a primary constructor parameter is shadowed? What happens when a base type has a protected member that is named differently than the parameter (ie, has an `_`)? This will make it easier to verify the behavior is what you're expecting, and to handle bugs as by-design or not.

> Also, for detecting the primary constructor, should check `constructor.Body == null && constructor.ExpressionBody == null`?

You only need the check that kimsey0 put in already. It's important to note that you can only detect primary constructors from source, you cannot detect them from metadata.

> good test as we don't support private fields.

Wait, what? I'll reiterate that I think you need spell out exactly what the rules are, in comments above the code, so that you can verify the code follows them. I thought I might understand the rules, but clearly I do not.

Reply:
@333fred, that sounds excellent. I'd be very happy to use a new `LookupSymbols` API for this. The pull request isn't exactly time sensitive either. I'd just be happy for it to land in .NET 9.

For now, I'll add the test cases you've suggested and try to find ways to handle them, if nothing else, then for the sake of my own learning.

Reply:
> Wait, what? I'll reiterate that I think you need spell out exactly what the rules are, in comments above the code, so that you can verify the code follows them. I thought I might understand the rules, but clearly I do not.

I am wrong here, looking more seeing we can support private fields https://learn.microsoft.com/en-us/dotnet/core/extensions/logger-message-generator#basic-usage this is because we generate the code inside the same class and LoggerMessage is not static. Sorry for confusion.

Reply:
> Should that discussion happen in this pull request or perhaps in https://github.com/dotnet/runtime/issues/91121?

Either way is fine. we can continue discussing in the PR and then put the conclusion in the issue. I am fine to move the discussion to the issue too. 

> Your comments above expose some edge cases, but there are probably more that I may not think of.

I think when applying the suggestion which looking up the fields first will address most these corner cases. Of course we'll need to document the behavior.

Reply:
I have added the three test cases suggested by @333fred, as well as one that shadows the primary constructor parameter with an `ILogger` field:

```
class Test(ILogger logger)
{
    private readonly ILogger logger = logger;
}
```

I've changed the code to support these cases by:
1. First looking for an `ILogger` field.
2. If none is found, look for an `ILogger` primary constructor parameter.
3. Ignore any primary constructor parameters which are shadowed by a field.

In practice, this means that:
1. Fields are prioritized over primary constructor parameters.
2. Shadowed primary constructor parameters are treated as if they don't exist.

I think this is a reasonable behavior, though one could argue that it would make sense to have a specific warning when a `ILogger` primary constructor parameter is found that is shadowed by a non-`ILogger` field.

How should this behavior be documented? Just on [the learn page](https://github.com/dotnet/docs/blob/main/docs/core/extensions/logger-message-generator.md)?

Reply:
> though one could argue that it would make sense to have a specific warning when a ILogger primary constructor parameter is found that is shadowed by a non-ILogger field.

Yes, we can have a diagnostic for that. Would it be better to have it as `Info` instead of `warning`? 

> How should this behavior be documented? Just on [the learn page](https://github.com/dotnet/docs/blob/main/docs/core/extensions/logger-message-generator.md)?

Right, this is the right place to document the behavior. We'll need to update this doc anyway for primary constructor usage with a sample.

Reply:
> Yes, we can have a diagnostic for that. Would it be better to have it as Info instead of warning?

I think that would make sense if we then still report the _no logger field found_ error in this case. I'll try implementing that.

> Right, this is the right place to document the behavior. We'll need to update this doc anyway for primary constructor usage with a sample.

Should I create a pull request over there before this merges, or is it fine to do after?

Reply:
It is fine we can add the doc after merging this PR.

Reply:
Sounds good. If both the behavior described above as well as the code implementing it is acceptable to all parties, I think this is ready for review again.

Reply:
I forgot to commit and push the info diagnostic, but just did that. I don't know if there are any guidelines for how exactly these diagnostic messages should be written or how the translation flow goes? (We can also just revert this. I think it's useful, but not necessary.)

Reply:
You don't have to worry about the translation of the added messages. I am seeing the added diagnostic is useful to catch the concerned case.